src/projectM-emscripten/projectM_SDL_emscripten.cpp

```
@@ -5,224 +5,275 @@
+void audioInputCallbackF32(void* userdata, unsigned char* stream, int len)
+{
+    if (app.audioChannelsCount == 1)
+    {
+        projectm_pcm_add_float_1ch_data(app.pm, reinterpret_cast<float*>(stream), len / sizeof(float));
+    }
+    else if (app.audioChannelsCount == 2)
+    {
+        projectm_pcm_add_float_2ch_data(app.pm, reinterpret_cast<float*>(stream), len / sizeof(float));
+    }
+    else
+    {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Multichannel audio not supported");
+    }
+}
+
+bool selectAudioInput(projectMApp* application)
+{
+    app.audioInputDevicesCount = SDL_GetNumAudioDevices(1);
 
-    if (! count) {
-        fprintf(stderr, "No audio input capture devices detected\n");
-        return 0;
+    if (!app.audioInputDevicesCount)
+    {
+        fprintf(stderr, "No audio input capture devices detected. Faking audio using random data.\n");
+        return false;
     }
 
-    printf("count: %d\n", count);
-    for (i = 0; i < count; ++i) {
-        printf("Audio device %d: %s\n", i, SDL_GetAudioDeviceName(i, 0));
+    printf("count: %d\n", app.audioInputDevicesCount);
+    for (int i = 0; i < app.audioInputDevicesCount; ++i)
+    {
+        printf("Audio device %d: %s\n", i, SDL_GetAudioDeviceName(i, 1));
     }
 
-    return 1;
+    SDL_AudioSpec want, have;
+
+    SDL_zero(want);
+    want.freq = 44100;
+    want.format = AUDIO_F32;
+    want.channels = 2;
+    want.samples = projectm_pcm_get_max_samples();
+    want.callback = &audioInputCallbackF32;
+
+    // Start with first device
+    app.audioInputDevice = SDL_OpenAudioDevice(SDL_GetAudioDeviceName(0, true), true, &want, &have, 0);
+    app.audioChannelsCount = have.channels;
+
+    if (have.channels == 1 || have.channels == 2)
+    {
+        SDL_PauseAudioDevice(app.audioInputDevice, false);
+        printf("Audio device specs: Channels=%d, Samplerate=%d, Format=%d\n", have.channels, have.freq, have.format);
+    }
```

maybe else explode?

---

Probably a case that never happens, so we're just not enabling the audio callback. Could alternatively fall back to the random audio generator to there's some movement at least.

In Chrome, Safari and Firefox at least, SDL always give 2-channel, 32-bit floating point audio with two channels, even if no input device is even connected. Looks more like a generic device that feeds from the Browser's HTML5 audio API.